components/encryption/src/master_key/kms.rs

```diff
@@ -148,92 +197,71 @@ where
+    encryption_backend: MemAesGcmBackend,
+    cached_ciphertext_key: Encrypted,
 }
 
-impl Inner {
-    fn maybe_update_backend(&mut self, ciphertext_key: Option<&Vec<u8>>) -> Result<()> {
-        let http_dispatcher = HttpClient::new().unwrap();
-        self.maybe_update_backend_with(ciphertext_key, http_dispatcher)
+impl State {
+    fn new_from_datakey(datakey: DataKey) -> Result<State> {
+        Ok(State {
+            cached_ciphertext_key: datakey.encrypted,
+            encryption_backend: MemAesGcmBackend::new(datakey.plaintext)?,
+        })
     }
 
-    fn maybe_update_backend_with<D>(
-        &mut self,
-        ciphertext_key: Option<&Vec<u8>>,
-        dispatcher: D,
-    ) -> Result<()>
-    where
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        if self.backend.is_some()
-            && ciphertext_key.map_or(true, |key| *key == self.cached_ciphertext_key)
-        {
-            return Ok(());
-        }
-
-        let mut kms = AwsKms::with_request_dispatcher(&self.config, dispatcher)?;
-        let key = if let Some(ciphertext_key) = ciphertext_key {
-            self.cached_ciphertext_key = ciphertext_key.to_owned();
-            kms.decrypt(ciphertext_key)?
-        } else {
-            let (ciphertext_key, plaintext_key) = kms.generate_data_key()?;
-            self.cached_ciphertext_key = ciphertext_key;
-            plaintext_key
-        };
-        if self.cached_ciphertext_key == key {
-            panic!(
-                "ciphertext key should not be the same as master key, \
-                otherwise it leaks master key!"
-            );
-        }
-
-        // Always use AES 256 for encrypting master key.
-        self.backend = Some(MemAesGcmBackend::new(key)?);
+    fn update_from_datakey(&mut self, datakey: DataKey) -> Result<()> {
+        self.cached_ciphertext_key = datakey.encrypted;
+        // Always use AES 256
+        self.encryption_backend = MemAesGcmBackend::new(datakey.plaintext)?;
         Ok(())
     }
+
+    fn cached(&self, ciphertext_key: &Encrypted) -> bool {
+        *ciphertext_key == self.cached_ciphertext_key
+    }
+}
+
+pub struct DataKey {
+    pub encrypted: Encrypted,
+    pub plaintext: Vec<u8>,
 }
 
+pub trait KmsProvider: Sync + Send + 'static + std::fmt::Debug {
+    fn generate_data_key(&mut self) -> Result<DataKey>;
+    fn decrypt(&mut self, ciphertext: EncryptedRef) -> Result<Vec<u8>>;
+    fn name(&self) -> &[u8];
+}
+
+#[derive(Debug)]
 pub struct KmsBackend {
-    inner: Mutex<Inner>,
+    state: Mutex<State>,
+    kms_provider: Mutex<Box<dyn KmsProvider>>,
 }
 
 impl KmsBackend {
-    pub fn new(config: KmsConfig) -> Result<KmsBackend> {
-        let inner = Inner {
-            backend: None,
-            config,
-            cached_ciphertext_key: Vec::new(),
-        };
-
+    pub fn new(mut kms_provider: Box<dyn KmsProvider>) -> Result<KmsBackend> {
+        let state = State::new_from_datakey(kms_provider.generate_data_key()?)?;
```

Comment:
Shall we keep the existing behavior to lazily initialize the kms data key? In case preexisting encryption metadata presents, we don't need this extra `generate_data_key` call.